go/benchmarks/benchmarking.go

```diff
@@ -0,0 +1,418 @@
+		clusterModeEnabled: runConfig.clusterModeEnabled,
+	}
+
+	err := executeBenchmarks(runConfig, connSettings)
+	if err != nil {
+		return err
+	}
+
+	if runConfig.resultsFile != os.Stdout {
+		return writeResults(runConfig.resultsFile)
+	}
+
+	return nil
+}
+
+type benchmarkConfig struct {
+	clientName         string
+	numConcurrentTasks int
+	clientCount        int
+	dataSize           int
+	minimal            bool
+	connectionSettings *connectionSettings
+	resultsFile        *os.File
+}
+
+func executeBenchmarks(runConfig *runConfiguration, connectionSettings *connectionSettings) error {
+	for _, clientName := range runConfig.clientNames {
+		for _, numConcurrentTasks := range runConfig.concurrentTasks {
+			for _, clientCount := range runConfig.clientCount {
+				for _, dataSize := range runConfig.dataSize {
+					benchmarkConfig := &benchmarkConfig{
+						clientName:         clientName,
+						numConcurrentTasks: numConcurrentTasks,
+						clientCount:        clientCount,
+						dataSize:           dataSize,
+						minimal:            runConfig.minimal,
+						connectionSettings: connectionSettings,
+						resultsFile:        runConfig.resultsFile,
+					}
+
+					err := runSingleBenchmark(benchmarkConfig)
+					if err != nil {
+						return err
+					}
+				}
+			}
+		}
+
+		fmt.Println()
+	}
+
+	return nil
+}
+
+func runSingleBenchmark(config *benchmarkConfig) error {
+	fmt.Printf("Running benchmarking for %s client:\n", config.clientName)
+	fmt.Printf(
+		"\n =====> %s <===== clientCount: %d, concurrentTasks: %d, dataSize: %d \n\n",
+		config.clientName,
+		config.clientCount,
+		config.numConcurrentTasks,
+		config.dataSize,
+	)
+
+	clients, err := createClients(config)
+	if err != nil {
+		return err
+	}
+
+	benchmarkResult := measureBenchmark(clients, config)
+	if config.resultsFile != os.Stdout {
+		addJsonResults(config, benchmarkResult)
+	}
+
+	printResults(benchmarkResult)
+	return closeClients(clients)
+}
+
+type benchmarkClient interface {
+	connect(connectionSettings *connectionSettings) error
+	set(key string, value string) (string, error)
+	get(key string) (string, error)
+	close() error
+	getName() string
+}
+
+func createClients(config *benchmarkConfig) ([]benchmarkClient, error) {
+	var clients []benchmarkClient
+	for clientNum := 0; clientNum < config.clientCount; clientNum++ {
+		var client benchmarkClient
+		switch config.clientName {
+		case goRedis:
+			client = &goRedisBenchmarkClient{}
+		case glide:
+			client = &glideBenchmarkClient{}
+		}
+
+		err := client.connect(config.connectionSettings)
+		if err != nil {
+			return nil, err
+		}
+
+		clients = append(clients, client)
+	}
+
+	return clients, nil
+}
+
+func closeClients(clients []benchmarkClient) error {
+	for _, client := range clients {
+		err := client.close()
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+var jsonResults []map[string]interface{}
+
+func writeResults(file *os.File) error {
+	encoder := json.NewEncoder(file)
```

Comment:
Do this extend existing file or overwrite it?

Reply:
overwrite

Reply:
Should extend, because benchmarking script calls bencmarking app multiple time with different parameters